### PR TITLE
Refactor OpenGL painter

### PR DESCRIPTION
**Pull request by toaster**
### Description:

This PR refactors the OpenGL painter implementation to reduce the duplicated code.
Since it is a refactoring without new features, there are no new tests.

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

File: internal/painter/gl/painter.go [hunk 1]
```diff
@@ -275,8 +275,5 @@ func (p *painter) textureScale(v float32) float32 {
 func NewPainter(c fyne.Canvas, ctx driver.WithContext) Painter {
 	p := &painter{canvas: c, contextProvider: ctx}
 	p.SetFrameBufferScale(1.0)
-
-	p.glInit()
```
Comment (andydotxyz) [inline internal/painter/gl/painter.go]: I may be missing something, but in this context it looks like some init code isn't being called any more - is this intended?

**Review by andydotxyz — COMMENTED** [review]
Great work.
Not sure about this removal of a `glInit` call - was it useless, or is it now called implicitly elsewhere?

File: internal/painter/gl/painter.go (hunk 1 above)
Comment (toaster) [reply internal/painter/gl/painter.go]: If “in this context” means “at this point of time”, yes. The code in question is executed later with `p.Init()`. The painter is useless unless `Init()` has been called anyhow, so no harm done. OTOH, the error logging in `glInit()` depends on the painter’s context been initialized which is done consistently in `p.Init()` for all implementations. At least for “mobile” it is not possible to do so earlier.

Comment (toaster) [reply internal/painter/gl/painter.go]: See also: 1fd952073c893b5fae88a0abf9eed845a3ce1ae1

Comment (andydotxyz) [reply internal/painter/gl/painter.go]: Thanks for the explanation, sounds like it's sorted :)

**Review by andydotxyz — APPROVED** [review]

**Comment by Bluebugs** [comment]
If you want to get the code coverage back up a bit, I think there is a few function in draw.go that can be tested fairly easily.

**Comment by toaster** [comment]
> If you want to get the code coverage back up a bit, I think there is a few function in draw.go that can be tested fairly easily.

Good idea, I’ll add a separate PR for this.

**Comment by changkun** [comment]
After merging this, the macOS/amd64 starts to crash at the start:

```
$ ./fyne_demo                                                                                                                                                      278a3824 ● [10:35:56]
2022/05/02 10:35:59 Lifecycle: Started
2022/05/02 10:35:59 Lifecycle: Entered Foreground
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc pc=0x7ffa2e7ec5a7]

runtime stack:
runtime.throw({0x45e1506?, 0x2e01fe00?})
        /Users/changkun/goes/go/src/runtime/panic.go:992 +0x71
runtime.sigpanic()
        /Users/changkun/goes/go/src/runtime/signal_unix.go:802 +0x3a9

goroutine 34 [syscall, locked to thread]:
runtime.cgocall(0x44d26e0, 0xc00072b5d0)
        /Users/changkun/goes/go/src/runtime/cgocall.go:157 +0x5c fp=0xc00072b5a8 sp=0xc00072b570 pc=0x4007b3c
github.com/go-gl/gl/v3.2-core/gl._Cfunc_glowDrawArrays(0x7ffa2e6a7583, 0x4, 0x0, 0x6)
        _cgo_gotypes.go:6459 +0x45 fp=0xc00072b5d0 sp=0xc00072b5a8 pc=0x431a905
github.com/go-gl/gl/v3.2-core/gl.DrawArrays(...)
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/go-gl/gl/v3.2-core/gl/package.go:9566
fyne.io/fyne/v2/internal/painter/gl.(*coreContext).DrawArrays(0xc00052aac0?, 0x45c2822?, 0x45c3bc7?, 0x18?)
        /Users/changkun/dev/changkun.de/fyne/internal/painter/gl/gl_core.go:156 +0x30 fp=0xc00072b5f8 sp=0xc00072b5d0 pc=0x43351f0
fyne.io/fyne/v2/internal/painter/gl.(*painter).drawLine(0xc00052aac0, 0xc000520690, {0x426b6fc?, 0x0?}, {0x4335445?, 0x0?})
        /Users/changkun/dev/changkun.de/fyne/internal/painter/gl/draw.go:85 +0x2f1 fp=0xc00072b660 sp=0xc00072b5f8 pc=0x43331d1
fyne.io/fyne/v2/internal/painter/gl.(*painter).drawObject(0xc00052aac0, {0x46c4b30, 0xc000520690}, {0x400f39d?, 0x0?}, {0x453afa0?, 0x0?})
        /Users/changkun/dev/changkun.de/fyne/internal/painter/gl/draw.go:139 +0xbf fp=0xc00072b6c0 sp=0xc00072b660 pc=0x4333a9f
fyne.io/fyne/v2/internal/painter/gl.(*painter).Paint(0x10051c680?, {0x46c4b30, 0xc000520690}, {0x400f39d?, 0x0?}, {0x453afa0?, 0x0?})
        /Users/changkun/dev/changkun.de/fyne/internal/painter/gl/painter.go:88 +0x74 fp=0xc00072b710 sp=0xc00072b6c0 pc=0x43363b4
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).paint.func1(0x426b131?, {0x426b6fc?, 0x0?})
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/canvas.go:287 +0x15f fp=0xc00072b778 sp=0xc00072b710 pc=0x4469ddf
fyne.io/fyne/v2/internal/driver/common.(*Canvas).walkTree.func1({0x46c4b30?, 0xc000520690}, {0x520690?, 0xc0?}, {0x52a300?, 0xc0?}, {0x52aac0?, 0xc0?})
        /Users/changkun/dev/changkun.de/fyne/internal/driver/common/canvas.go:438 +0x277 fp=0xc00072b7d0 sp=0xc00072b778 pc=0x433a937
fyne.io/fyne/v2/internal/driver.walkObjectTree({0x46c4b30, 0xc000520690}, 0x0, {0x46c4a10, 0xc00052a300}, {0x453afa0?, 0x0?}, {0x44620000?, 0x44200000?}, {0x433c5800?, ...}, ...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:169 +0x29b fp=0xc00072b898 sp=0xc00072b7d0 pc=0x42622bb
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree({0x46c4a10, 0xc00052a300}, 0x0, {0x46c4a10, 0xc00052a0c0}, {0x453afa0?, 0x0?}, {0x44620000?, 0x44200000?}, {0x433c5800?, ...}, ...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:190 +0x41d fp=0xc00072b960 sp=0xc00072b898 pc=0x426243d
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree({0x46c4a10, 0xc00052a0c0}, 0x0, {0x46c4a10, 0xc00052a240}, {0x4cc3300?, 0x0?}, {0x44620000?, 0x44200000?}, {0x40800000?, ...}, ...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:190 +0x41d fp=0xc00072ba28 sp=0xc00072b960 pc=0x426243d
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree({0x46c4a10, 0xc00052a240}, 0x0, {0x46c50d0, 0xc0003e09a0}, {0x0?, 0x0?}, {0x0?, 0x0?}, {0x0?, ...}, ...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:190 +0x41d fp=0xc00072baf0 sp=0xc00072ba28 pc=0x426243d
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree({0x46c50d0, 0xc0003e09a0}, 0x0, {0x0, 0x0}, {0x0?, 0x0?}, {0xafc30?, 0xc0?}, {0x40418b1?, ...}, ...)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:190 +0x41d fp=0xc00072bbb8 sp=0xc00072baf0 pc=0x426243d
fyne.io/fyne/v2/internal/driver.WalkVisibleObjectTree({0x46c50d0?, 0xc0003e09a0?}, 0xc0000afc60?, 0x4061cbb?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/util.go:134 +0x4c fp=0xc00072bc20 sp=0xc00072bbb8 pc=0x4261fcc
fyne.io/fyne/v2/internal/driver/common.(*Canvas).walkTree(0xc000000000?, 0xc00050efa0, 0xc0000afd80, 0xc0000afd68)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/common/canvas.go:459 +0x165 fp=0xc00072bcd8 sp=0xc00072bc20 pc=0x433a525
fyne.io/fyne/v2/internal/driver/common.(*Canvas).WalkTrees(0xc000524000, 0xc0000afd50?, 0x44694e6?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/common/canvas.go:371 +0x3c fp=0xc00072bd30 sp=0xc00072bcd8 pc=0x433a07c
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).paint(0xc000524000, {0x0?, 0x0?})
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/canvas.go:301 +0xfe fp=0xc00072bdb0 sp=0xc00072bd30 pc=0x4469b5e
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).repaintWindow.func1()
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:202 +0x8f fp=0xc00072be00 sp=0xc00072bdb0 pc=0x446c72f
fyne.io/fyne/v2/internal/driver/glfw.(*window).RunWithContext(0xc0003e0620?, 0xc00017fe30)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/window.go:903 +0x4f fp=0xc00072be20 sp=0xc00072be00 pc=0x447474f
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).repaintWindow(0xc00017fe98?, 0xc000414340?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:193 +0x4a fp=0xc00072be58 sp=0xc00072be20 pc=0x446c66a
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).drawSingleFrame(0xc00017ff90?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:101 +0x1cb fp=0xc00072bef8 sp=0xc00072be58 pc=0x446be4b
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).startDrawThread.func1()
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:249 +0x1a6 fp=0xc00072bfe0 sp=0xc00072bef8 pc=0x446cba6
runtime.goexit()
        /Users/changkun/goes/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc00072bfe8 sp=0xc00072bfe0 pc=0x4066961
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).startDrawThread
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:225 +0xea

goroutine 1 [syscall, locked to thread]:
github.com/go-gl/glfw/v3.3/glfw._Cfunc_glfwPollEvents()
        _cgo_gotypes.go:1477 +0x45
github.com/go-gl/glfw/v3.3/glfw.PollEvents()
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/go-gl/glfw/v3.3/glfw/window.go:964 +0x17
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).tryPollEvents(0xc00072fe38?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop_desktop.go:39 +0x3b
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).runGL(0xc0003e0620)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/loop.go:133 +0x1f9
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).Run(0xc000518000?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/driver.go:156 +0x30
fyne.io/fyne/v2/internal/driver/glfw.(*window).ShowAndRun(0xc000518000)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/window.go:230 +0x2c
main.main()
        /Users/changkun/dev/changkun.de/fyne/cmd/fyne_demo/main.go:67 +0x48c

goroutine 6 [chan receive]:
fyne.io/fyne/v2/test.NewApp.func1()
        /Users/changkun/dev/changkun.de/fyne/test/testapp.go:117 +0x45
created by fyne.io/fyne/v2/test.NewApp
        /Users/changkun/dev/changkun.de/fyne/test/testapp.go:115 +0x34d

goroutine 7 [syscall]:
syscall.syscall6(0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/changkun/goes/go/src/runtime/sys_darwin.go:44 +0x6a
golang.org/x/sys/unix.kevent(0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/golang.org/x/sys/unix/zsyscall_darwin_amd64.go:276 +0x52
golang.org/x/sys/unix.Kevent(0x0?, {0x0?, 0x0?, 0x0?}, {0xc00005ce78?, 0x400b0c5?, 0x0?}, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/golang.org/x/sys/unix/syscall_bsd.go:429 +0x3c
github.com/fsnotify/fsnotify.read(0x0?, {0xc00005ce78?, 0x0?, 0xa}, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:512 +0x3f
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0000386c0)
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:275 +0xa5
created by github.com/fsnotify/fsnotify.NewWatcher
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:63 +0x1c5

goroutine 8 [chan receive]:
fyne.io/fyne/v2/app.watchFile.func1()
        /Users/changkun/dev/changkun.de/fyne/app/settings_desktop.go:43 +0x6c
created by fyne.io/fyne/v2/app.watchFile
        /Users/changkun/dev/changkun.de/fyne/app/settings_desktop.go:42 +0x10a

goroutine 18 [syscall]:
syscall.syscall6(0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/changkun/goes/go/src/runtime/sys_darwin.go:44 +0x6a
golang.org/x/sys/unix.kevent(0x0?, 0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/golang.org/x/sys/unix/zsyscall_darwin_amd64.go:276 +0x52
golang.org/x/sys/unix.Kevent(0x0?, {0x0?, 0x0?, 0x0?}, {0xc00005de78?, 0x400b0c5?, 0x0?}, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/golang.org/x/sys/unix/syscall_bsd.go:429 +0x3c
github.com/fsnotify/fsnotify.read(0x0?, {0xc00005de78?, 0x0?, 0xa}, 0x0?)
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:512 +0x3f
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0004140c0)
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:275 +0xa5
created by github.com/fsnotify/fsnotify.NewWatcher
        /Users/changkun/dev/changkun.de/fyne/vendor/github.com/fsnotify/fsnotify/kqueue.go:63 +0x1c5

goroutine 19 [chan receive]:
fyne.io/fyne/v2/app.watchFile.func1()
        /Users/changkun/dev/changkun.de/fyne/app/settings_desktop.go:43 +0x6c
created by fyne.io/fyne/v2/app.watchFile
        /Users/changkun/dev/changkun.de/fyne/app/settings_desktop.go:42 +0x10a

goroutine 35 [select]:
fyne.io/fyne/v2/internal/async.(*UnboundedFuncChan).processing(0xc000520000)
        /Users/changkun/dev/changkun.de/fyne/internal/async/chan_func.go:45 +0xda
created by fyne.io/fyne/v2/internal/async.NewUnboundedFuncChan
        /Users/changkun/dev/changkun.de/fyne/internal/async/chan_func.go:22 +0xf8

goroutine 36 [chan receive]:
fyne.io/fyne/v2/internal/driver/common.(*Window).RunEventQueue(0x0?)
        /Users/changkun/dev/changkun.de/fyne/internal/driver/common/window.go:34 +0x4d
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).createWindow.func1
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/window.go:942 +0x136

goroutine 10 [sleep]:
time.Sleep(0x3b9aca00)
        /Users/changkun/goes/go/src/runtime/time.go:194 +0x12e
fyne.io/fyne/v2/cmd/fyne_demo/tutorials.canvasScreen.func1()
        /Users/changkun/dev/changkun.de/fyne/cmd/fyne_demo/tutorials/canvas.go:25 +0x35
created by fyne.io/fyne/v2/cmd/fyne_demo/tutorials.canvasScreen
        /Users/changkun/dev/changkun.de/fyne/cmd/fyne_demo/tutorials/canvas.go:23 +0x136

goroutine 66 [chan receive]:
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).SetSystemTrayMenu.func1.1()
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/driver_desktop.go:60 +0x45
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).SetSystemTrayMenu.func1
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/driver_desktop.go:59 +0x1ca

goroutine 67 [chan receive]:
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).SetSystemTrayMenu.func1.2()
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/driver_desktop.go:69 +0x2b
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).SetSystemTrayMenu.func1
        /Users/changkun/dev/changkun.de/fyne/internal/driver/glfw/driver_desktop.go:68 +0x29e
```

**Comment by Jacalz** [comment]
Yeah. I can replicate the same behaviour on linux/amd64.

**Comment by andydotxyz** [comment]
This is very peculiar as neither of those platforms is erring for me...

**Comment by Bluebugs** [comment]
I am seeing this crash too when using delv on Linux. The crash is in drawLine in gl/draw.go on the DrawArrays call. I do not understand why as the DrawArrays in drawTextureWithDetails work just fine.

**Comment by andydotxyz** [comment]
Ah yes, the Canvas tab on fyne_demo shows it!

**Comment by andydotxyz** [comment]
See #2962

**Pull request merged**